core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1298,18 +1298,18 @@ class SizeBasedCoalescer(val maxSize: Int) extends PartitionCoalescer with Seria
       val splitSize = fileSplit.getLength
       if (currentSum + splitSize < maxSize) {
         addPartition(partition, splitSize)
-        index += 1
-        if (index == partitions.size) {
-          updateGroups
-        }
       } else {
-        if (currentGroup.partitions.size == 0) {
+        if (currentGroup.partitions.isEmpty) {
           addPartition(partition, splitSize)
-          index += 1
         } else {
-          updateGroups
+          updateGroups()
+          addPartition(partition, splitSize)
         }
       }
+      index += 1
+      if (index == partitions.length) {
+        updateGroups()
+      }
```

Comment:
BTW, the previous code was added at 2.0.0. Did you hit a test flakiness due to this?

Reply:
> BTW, the previous code was added at 2.0.0. Did you hit a test flakiness due to this?

Yeah, these days I am testing a way to  fix spark small out put file problem, and I use this logic, during test, found that this logic is wrong.